RaftMongoWithRaftReconfig.tla

```diff
@@ -0,0 +1,276 @@
+vars == <<serverVars, logVars, configs>>
+
+----
+\* Helpers
+
+\* The term of the last entry in a log, or 0 if the log is empty.
+GetTerm(xlog, index) == IF index = 0 THEN 0 ELSE xlog[index].term
+LogTerm(i, index) == GetTerm(log[i], index)
+LastTerm(xlog) == GetTerm(xlog, Len(xlog))
+
+\* Return the minimum value from a set, or undefined if the set is empty.
+Min(s) == CHOOSE x \in s : \A y \in s : x <= y
+\* Return the maximum value from a set, or undefined if the set is empty.
+Max(s) == CHOOSE x \in s : \A y \in s : x >= y
+
+\* The config version in the node's last entry.
+GetConfigVersion(i) == log[i][Len(log[i])].configVersion
+
+\* Gets the node's first entry with a given config version.
+GetConfigEntry(i, configVersion) == LET configEntries == {index \in 1..Len(log[i]) : 
+                                                            log[i][index].configVersion = configVersion}
+                                    IN Min(configEntries)
+
+\* The servers that are in the same config as i.
+ServerViewOn(i) == configs[GetConfigVersion(i)]
+
+\* The set of all quorums. This just calculates simple majorities, but the only
+\* important property is that every quorum overlaps with every other.
+Quorum(me) == {sub \in SUBSET(ServerViewOn(me)) : Cardinality(sub) * 2 > Cardinality(ServerViewOn(me))}
+
+----
+\* Define initial values for all variables
+InitServerVars == /\ currentTerm = [i \in Server |-> 0]
+                  /\ state       = [i \in Server |-> Follower]
+                  /\ commitPoint = [i \in Server |-> [term |-> 0, index |-> 0]]
+InitLogVars == /\ log         = [i \in Server |-> << [term |-> 0, configVersion |-> 1] >>]
+InitConfigs == configs = << Server >>
+Init == /\ InitServerVars
+        /\ InitLogVars
+        /\ InitConfigs
+
+----
+\* Message handlers
+\* i = recipient, j = sender, m = message
+
+AppendOplog(i, j) ==
+    /\ state[i] = Follower  \* Disable primary catchup and draining
+    /\ j \in ServerViewOn(i)  \* j is in the config of i.
+    /\ Len(log[i]) < Len(log[j])
+    /\ LastTerm(log[i]) = LogTerm(j, Len(log[i]))
+    /\ log' = [log EXCEPT ![i] = Append(log[i], log[j][Len(log[i]) + 1])]
+    /\ currentTerm' = [currentTerm EXCEPT ![i] = IF LogTerm(j, Len(log[i]) + 1) > currentTerm[i] 
+                                                 THEN LogTerm(j, Len(log[i]) + 1) 
+                                                 ELSE currentTerm[i]]
+    /\ UNCHANGED <<state, commitPoint, configs>>
+
+CanRollbackOplog(i, j) ==
+    /\ j \in ServerViewOn(i)  \* j is in the config of i.
+    /\ Len(log[i]) > 0
+    /\ \* The log with later term is more up-to-date
+       LastTerm(log[i]) < LastTerm(log[j])
+    /\
+       \/ Len(log[i]) > Len(log[j])
+       \* There seems no short-cut of OR clauses, so I have to specify the negative case
+       \/ /\ Len(log[i]) <= Len(log[j])
+          /\ LastTerm(log[i]) /= LogTerm(j, Len(log[i]))
+
+RollbackOplog(i, j) ==
```

Comment:
I still feel confused. How could a node magically know that a certain set of entries were uncommitted? I think the reason this is correct is because even this happens after the voting, it happens to be equivalent to rollback before the voting.

Reply:
I prefer to leave out the term updates here because they are not necessary for safety, and I want the spec to only include the term updates that are necessary for safety.